### PR TITLE
fix indices for data_parallel and add parameter gradient tests

### LINKED ISSUE #2612: Multi-GPU tarining with DataParallel must has GPU0
The loss will never decrease when I train a model using multi-gpus but without gpu0. This is an example I modified from the [official mnist example](https://github.com/pytorch/examples/tree/master/mnist) : https://gist.github.com/longcw/e5efdb0e5a580a6b75604561b6e4f613

```python
gpus = [1, 2]
model = torch.nn.DataParallel(model, device_ids=gpus)
```

If `gpus = [1, 2]` the result after one epoch is 
```
...
Train Epoch: 1 [58240/60000 (97%)]      Loss: 2.363280
Train Epoch: 1 [58880/60000 (98%)]      Loss: 2.373188
Train Epoch: 1 [59520/60000 (99%)]      Loss: 2.351889

Test set: Average loss: 2.3285, Accuracy: 685/10000 (7%)
```

If `gpus = [0, 2]` the result after one epoch is 
```
...
Train Epoch: 1 [58240/60000 (97%)]	Loss: 0.268342
Train Epoch: 1 [58880/60000 (98%)]	Loss: 0.286652
Train Epoch: 1 [59520/60000 (99%)]	Loss: 0.514245

Test set: Average loss: 0.1972, Accuracy: 9436/10000 (94%)
```

### PR DESCRIPTION
**Pull request by ngimel**
Hopefully, should fix #2612 and #2573

### CODE REVIEW

**Review by soumith — APPROVED** [review]

File: torch/cuda/comm.py [hunk 1]
```diff
@@ -74,22 +74,24 @@ def reduce_add(inputs, destination=None):
     """
     # TODO: try to find an input on another gpu, copy it,
     # and accumulate into the copy
+    if destination is None:
+        destination = torch.cuda.current_device()
     input_size = inputs[0].size()
     for i, inp in enumerate(inputs):
         assert inp.is_cuda, "reduce_add expects all inputs to be on GPUs"
+        if inp.get_device() == destination:
+            nccl_root = i
```
Comment (apaszke) [inline torch/cuda/comm.py]: The problem with this is that it's ambiguous if there are two tensors on destination device. Can you add `assert nccl_root is None` + `assert nccl_root is not None` to check that we really found the root after the loop?

Comment (ngimel) [reply torch/cuda/comm.py]: I can add asserts. nccl2 is actually removing support for the cases when a few tensors are on the same device, and I don't think you can have root that is not the device of one of the tensors.

**Comment by apaszke** [comment]
The linter still complains about the tests
